subsys/bluetooth/services/dis.c

```diff
@@ -45,105 +47,231 @@ static struct dis_pnp dis_pnp_id = {
 
-#define BT_DIS_MODEL_REF		CONFIG_BT_DIS_MODEL
-#define BT_DIS_MANUF_REF		CONFIG_BT_DIS_MANUF
-#define BT_DIS_SERIAL_NUMBER_STR_REF	CONFIG_BT_DIS_SERIAL_NUMBER_STR
-#define BT_DIS_FW_REV_STR_REF		CONFIG_BT_DIS_FW_REV_STR
-#define BT_DIS_HW_REV_STR_REF		CONFIG_BT_DIS_HW_REV_STR
-#define BT_DIS_SW_REV_STR_REF		CONFIG_BT_DIS_SW_REV_STR
+#define BT_DIS_MODEL_REF             CONFIG_BT_DIS_MODEL
+#define BT_DIS_MANUF_REF             CONFIG_BT_DIS_MANUF
+#define BT_DIS_SERIAL_NUMBER_STR_REF CONFIG_BT_DIS_SERIAL_NUMBER_STR
+#define BT_DIS_FW_REV_STR_REF        CONFIG_BT_DIS_FW_REV_STR
+#define BT_DIS_HW_REV_STR_REF        CONFIG_BT_DIS_HW_REV_STR
+#define BT_DIS_SW_REV_STR_REF        CONFIG_BT_DIS_SW_REV_STR
+#define BT_DIS_UDI_LABEL_STR_REF     CONFIG_BT_DIS_UDI_LABEL_STR
+#define BT_DIS_UDI_DI_STR_REF        CONFIG_BT_DIS_UDI_DI_STR
+#define BT_DIS_UDI_ISSUER_STR_REF    CONFIG_BT_DIS_UDI_ISSUER_STR
+#define BT_DIS_UDI_AUTHORITY_STR_REF CONFIG_BT_DIS_UDI_AUTHORITY_STR
+#define BT_DIS_IEEE_RCDL_STR_REF     CONFIG_BT_DIS_IEEE_RCDL_STR
+
+#define BT_DIS_STRLEN(x) strlen(x)
 
 #endif /* CONFIG_BT_DIS_SETTINGS */
 
-static ssize_t read_str(struct bt_conn *conn,
-			  const struct bt_gatt_attr *attr, void *buf,
-			  uint16_t len, uint16_t offset)
+static ssize_t read_str(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			uint16_t len, uint16_t offset)
 {
 	return bt_gatt_attr_read(conn, attr, buf, len, offset, attr->user_data,
-				 strlen(attr->user_data));
+				 BT_DIS_STRLEN(attr->user_data));
 }
 
 #if CONFIG_BT_DIS_PNP
-static ssize_t read_pnp_id(struct bt_conn *conn,
-			   const struct bt_gatt_attr *attr, void *buf,
+static ssize_t read_pnp_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
 			   uint16_t len, uint16_t offset)
 {
-	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id,
-				 sizeof(dis_pnp_id));
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, &dis_pnp_id, sizeof(dis_pnp_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_SYSTEM_ID
+static ssize_t read_system_id(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			      uint16_t len, uint16_t offset)
+{
+	return bt_gatt_attr_read(conn, attr, buf, len, offset, dis_system_id,
+				 sizeof(dis_system_id));
+}
+#endif
+
+#if CONFIG_BT_DIS_UDI
+#define BT_DIS_UDI_FLAG_LABEL     (!!BT_DIS_UDI_LABEL_STR_REF[0])
+#define BT_DIS_UDI_FLAG_DI        (!!BT_DIS_UDI_DI_STR_REF[0])
+#define BT_DIS_UDI_FLAG_ISSUER    (!!BT_DIS_UDI_ISSUER_STR_REF[0])
+#define BT_DIS_UDI_FLAG_AUTHORITY (!!BT_DIS_UDI_AUTHORITY_STR_REF[0])
+#define BT_DIS_UDI_FLAGS                                                                           \
+	(BT_DIS_UDI_FLAG_LABEL | (BT_DIS_UDI_FLAG_DI << 1) | (BT_DIS_UDI_FLAG_ISSUER << 2) |       \
+	 (BT_DIS_UDI_FLAG_AUTHORITY << 3))
+
+static ssize_t read_udi(struct bt_conn *conn, const struct bt_gatt_attr *attr, void *buf,
+			uint16_t len, uint16_t offset)
+{
+	uint16_t written_bytes = 0;
+	uint16_t index = offset;
```

Comment:
Suggest to add some comments for this function as it may be non-obvious to most why this specific string is treated significantly different

Reply:
Done